sycl/doc/design/ThinLTO.md

```diff
@@ -0,0 +1,149 @@
+
+With traditional SYCL device code linking, all user code is linked together
+along with device libraries into a single huge module and then split and
+processed by `sycl-post-link`. This requires sequential processing, has a large
+memory footprint, and differs from the linking flow for AMD and NVIDIA devices.
+
+## Summary
+
+SYCL ThinLTO will hook into the existing community mechanism to run LTO as part
+of device linking inside `clang-linker-wrapper`. We split the device images
+early at compilation time, and at link time we use ThinLTO's function importing
+feature to bring in the definitions for referenced functions. Only the new
+offload model is supported.
+
+## Device code compilation time changes
+
+Most of the changes for ThinLTO occur during device link time, however there is
+one major change during compilation (-c) time: we now run device code split
+during compilation instead of linking. The main reason for doing this is
+increased parallelization. Many compilation jobs can be run at the same time,
+but linking happens once total for the application. Device code split is
+currently a common source of performance issues.
+
+Splitting early means that the resulting IR after splitting is not complete, it
+still may contain calls to functions (user code and/or the SYCL device
+libraries) defined in other translation units.
+
+We rely on the assumption that all function definitions matching a declaration
+will be the same and we can let ThinLTO pull in any one.
+
+For example, let's start with user device code that defines a `SYCL_EXTERNAL`
+function `foo` in translation unit `tu_foo`. There is also another translation
+unit `tu_bar` that references `foo`. During the early device code splitting run
+of `tu_foo`, we may find that more than one of the resultant device images
+contain a definition for `foo`.
+
+We assert that any function definition for `foo` that is deemed a match by the
+ThinLTO infrastructure during the processing of `tu_bar` is valid.
+
+As a result of running early device code split, the fat object file generated as
+part of device compilation may contain multiple device code images.
+
+## Device code link time changes
+
+Before we go into the link time changes for SYCL, let's understand the device
+linking flow for AMD/NVIDIA devices:
+
+![Community linking flow](images/ThinLTOCommunityFlow.svg)
+
+SYCL has two differentiating requirements:
+
+1) The SPIR-V backend is not production ready and the SPIR-V translator is used.
+2) The SYCL runtime requires metadata (module properties and module symbol
+table) computed from device images that will be stored along the device images
+in the fat executable.
+
+The effect of requirement 1) is that instead of letting ThinLTO call the SPIR-V
+backend, we add a callback that runs right before CodeGen would run. In that
+callback, we call the SPIR-V translator and store the resultant file path for
+use later, and we instruct the ThinLTO framework to not perform CodeGen.
```

Comment:
Note: we should be able to call the translator as a library and store the final binary (in form of a `MemoryBuffer`, I assume) instead of file path. The former is supposedly more in line with what happens in the upstream.